staging/src/k8s.io/apiserver/pkg/util/peerproxy/peerproxy.go

```diff
@@ -35,33 +42,46 @@ type Interface interface {
 }
 
 // New creates a new instance to implement unknown version proxy
+// This method is used for an alpha feature UnknownVersionInteroperabilityProxy
+// and is subject to future modifications.
 func NewPeerProxyHandler(informerFactory kubeinformers.SharedInformerFactory,
-	svm storageversion.Manager,
-	proxyTransport http.RoundTripper,
 	serverId string,
 	reconciler reconcilers.PeerEndpointLeaseReconciler,
-	serializer runtime.NegotiatedSerializer) *peerProxyHandler {
+	ser runtime.NegotiatedSerializer,
+	loopbackClientConfig *rest.Config,
+	proxyClientConfig *transport.Config) (*peerProxyHandler, error) {
 	h := &peerProxyHandler{
-		name:                  "PeerProxyHandler",
-		storageversionManager: svm,
-		proxyTransport:        proxyTransport,
-		svMap:                 sync.Map{},
-		serverId:              serverId,
-		reconciler:            reconciler,
-		serializer:            serializer,
+		name:                          "PeerProxyHandler",
+		serverId:                      serverId,
+		reconciler:                    reconciler,
+		serializer:                    ser,
+		loopbackClientConfig:          loopbackClientConfig,
+		proxyClientConfig:             proxyClientConfig,
+		localDiscoveryResponseCache:   make(map[schema.GroupVersion][]string),
+		localDiscoveryCacheTicker:     time.NewTicker(30 * time.Minute),
+		peerAggDiscoveryResponseCache: make(map[string]*peerAggDiscoveryInfo),
 	}
-	svi := informerFactory.Internal().V1alpha1().StorageVersions()
-	h.storageversionInformer = svi.Informer()
 
-	svi.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+	h.apiserverIdentityInformer = informerFactory.Coordination().V1().Leases().Informer()
+	h.apiserverIdentityLister = informerFactory.Coordination().V1().Leases().Lister()
+	discoveryScheme := runtime.NewScheme()
+	utilruntime.Must(apidiscoveryv2.AddToScheme(discoveryScheme))
+	h.discoverySerializer = serializer.NewCodecFactory(discoveryScheme)
+
+	_, err := h.apiserverIdentityInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
 		AddFunc: func(obj interface{}) {
-			h.addSV(obj)
+			h.addPeerDiscoveryInfo(obj)
 		},
 		UpdateFunc: func(oldObj, newObj interface{}) {
-			h.updateSV(oldObj, newObj)
+			h.updatePeerDiscoveryInfo(oldObj, newObj)
```

Comment:
These functions could potentially take a while, it's probably better to use a queue like in other controllers.